File: oxidize-python/oxidize_python/core/model/llama_decoder_forward.py
"""Decoder forward pass mirroring oxidize-golang/core/model/llama_decoder_forward.go."""

from __future__ import annotations

import math
from collections.abc import Callable

from oxidize_python.core.flash_attention.flash_attention import flash_attention_decode_heads_gqa
from oxidize_python.core.model.llama_decoder import LlamaDecoderStack
from oxidize_python.core.model.model import Logits
from oxidize_python.core.tensor.ops import apply_rope_head_f32, rms_norm_f32


def _apply_glu(gate: list[float], up: list[float], gelu: bool) -> None:
    """In place: gate[i] = activation(gate[i]) * up[i]; GeGLU if gelu else SwiGLU."""
    n = min(len(gate), len(up))
    if gelu:
        k = 0.7978845608  # sqrt(2/pi)
        for i in range(n):
            g = gate[i]
            act = 0.5 * g * (1.0 + math.tanh(k * (g + 0.044715 * g * g * g)))
            gate[i] = act * up[i]
        return
    for i in range(n):
        g = gate[i]
        gate[i] = (g / (1.0 + math.exp(-g))) * up[i]


def _windowed_cache(
    keys: list[float], values: list[float], seq_len: int, kv_len: int, window: int
) -> tuple[list[float], list[float], int]:
    """Slice the cache to the most recent `window` rows (sliding-window attention).

    RoPE encodes absolute positions, so slicing off the oldest rows yields the
    windowed-causal mask with relative positions preserved.
    """
    if window <= 0 or seq_len <= window:
        return keys, values, seq_len
    skip = (seq_len - window) * kv_len
    return keys[skip:], values[skip:], window


def forward_token(stack: LlamaDecoderStack, token: int) -> list[float]:
    return forward_token_with_context(stack, token, None, None)


def forward_token_with_context(
    stack: LlamaDecoderStack,
    token: int,
    target_context: list[float] | None,
    kv_context: Callable[[int, list[float]], tuple[list[float], list[float]]] | None,
) -> list[float]:
    h = stack.config.hidden_size
    hidden = [0.0] * h
    stack.fill_token_embedding(token, hidden)
    stack.scale_embedding(hidden)

    for layer_idx, layer in enumerate(stack.layers):
        head_dim = stack.config.kv_head_dim()
        if layer.attention.q_norm_weight:
            head_dim = len(layer.attention.q_norm_weight)
        num_heads = stack.config.num_attention_heads
        num_kv = stack.config.num_key_value_heads
        q_size = num_heads * head_dim
        kv_len = num_kv * head_dim

        attn_out = [0.0] * q_size
        mlp_out = [0.0] * h

        normed = [0.0] * h
        rms_norm_f32(hidden, layer.input_layernorm, normed, stack.config.rms_norm_eps)

        q = [0.0] * q_size
        k = [0.0] * kv_len
        v = [0.0] * kv_len
        if layer.attention.q_proj.is_loaded():
            layer.attention.q_proj.gemv(normed, q)
        if layer.attention.k_proj.is_loaded():
            layer.attention.k_proj.gemv(normed, k)
        if layer.attention.v_proj.is_loaded():
            layer.attention.v_proj.gemv(normed, v)

        k_ctx: list[float] | None = None
        v_ctx: list[float] | None = None
        if kv_context is not None and target_context is not None:
            k_ctx, v_ctx = kv_context(layer_idx, target_context)

        head_scratch = [0.0] * head_dim
        pos = stack.position_offset

        def apply_qk_norm(vec: list[float], weight: list[float], heads: int) -> None:
            if len(weight) != head_dim:
                return
            for hi in range(heads):
                start = hi * head_dim
                rms_norm_f32(
                    vec[start : start + head_dim],
                    weight,
                    head_scratch,
                    stack.config.rms_norm_eps,
                )
                vec[start : start + head_dim] = head_scratch[:]

        apply_qk_norm(q, layer.attention.q_norm_weight, num_heads)
        apply_qk_norm(k, layer.attention.k_norm_weight, num_kv)
        if k_ctx is not None:
            apply_qk_norm(k_ctx, layer.attention.k_norm_weight, num_kv)

        layer_rope = stack.config.layer_rope_theta(layer_idx)

        def apply_rope(vec: list[float], heads: int, theta: float = layer_rope) -> None:
            for hi in range(heads):
                start = hi * head_dim
                apply_rope_head_f32(
                    vec[start : start + head_dim],
                    head_scratch,
                    pos,
                    head_dim,
                    theta,
                )
                vec[start : start + head_dim] = head_scratch[:]

        apply_rope(q, num_heads)
        apply_rope(k, num_kv)
        if k_ctx is not None:
            apply_rope(k_ctx, num_kv)

        cache = stack.kv_cache[layer_idx]
        if k_ctx is not None and v_ctx is not None:
            cache.keys.extend(k_ctx)
            cache.values.extend(v_ctx)
            cache.seq_len += 1
        cache.keys.extend(k)
        cache.values.extend(v)
        cache.seq_len += 1

        window = stack.config.layer_sliding_window(layer_idx)
        win_keys, win_values, win_seq = _windowed_cache(
            cache.keys, cache.values, cache.seq_len, kv_len, window
        )
        flash_attention_decode_heads_gqa(
            q,
            win_keys,
            win_values,
            attn_out,
            win_seq,
            head_dim,
            kv_len,
            num_heads,
            num_kv,
        )

        o_result = [0.0] * h
        if layer.attention.o_proj.is_loaded():
            layer.attention.o_proj.gemv(attn_out, o_result)
        else:
            o_result[: min(h, q_size)] = attn_out[: min(h, q_size)]
            for i in range(q_size, h):
                o_result[i] = 0.0
        attn_out = o_result

        # Gemma sandwich norm: normalize the attention output before residual.
        if stack.config.sandwich_norm and len(layer.post_attention_layernorm) == h:
            normed_attn = [0.0] * h
            rms_norm_f32(
                attn_out, layer.post_attention_layernorm, normed_attn, stack.config.rms_norm_eps
            )
            attn_out = normed_attn

        for i in range(h):
            hidden[i] += attn_out[i]

        mlp_norm = (
            layer.pre_ffn_layernorm
            if stack.config.sandwich_norm and layer.pre_ffn_layernorm
            else layer.post_attention_layernorm
        )
        normed = [0.0] * h
        rms_norm_f32(hidden, mlp_norm, normed, stack.config.rms_norm_eps)
        inter = stack.config.intermediate_size
        gate = [0.0] * inter
        up = [0.0] * inter
        if layer.mlp_gate.is_loaded():
            layer.mlp_gate.gemv(normed, gate)
        if layer.mlp_up.is_loaded():
            layer.mlp_up.gemv(normed, up)
        _apply_glu(gate, up, stack.config.gelu_ffn)
        if layer.mlp_down.is_loaded():
            layer.mlp_down.gemv(gate, mlp_out)

        # Gemma sandwich norm: normalize the FFN output before residual.
        if stack.config.sandwich_norm and len(layer.post_ffn_layernorm) == h:
            normed_ffn = [0.0] * h
            rms_norm_f32(mlp_out, layer.post_ffn_layernorm, normed_ffn, stack.config.rms_norm_eps)
            mlp_out = normed_ffn

        for i in range(h):
            hidden[i] += mlp_out[i]

    if stack.norm:
        out = [0.0] * h
        rms_norm_f32(hidden, stack.norm, out, stack.config.rms_norm_eps)
        hidden = out

    stack.position_offset += 1
    return hidden


def forward_batch(stack: LlamaDecoderStack, tokens: list[int]) -> list[float]:
    if not tokens:
        raise ValueError("empty token batch")
    if len(tokens) == 1:
        return forward_token(stack, tokens[0])

    b = len(tokens)
    h = stack.config.hidden_size
    hidden = [0.0] * (b * h)

    scale = stack.config.embedding_scale
    if stack.tok_embeddings.is_loaded():
        for t, tok in enumerate(tokens):
            row = [0.0] * h
            stack.fill_token_embedding(tok, row)
            if scale not in (0.0, 1.0):
                row = [v * scale for v in row]
            hidden[t * h : (t + 1) * h] = row

    for layer_idx, layer in enumerate(stack.layers):
        head_dim = stack.config.kv_head_dim()
        if layer.attention.q_norm_weight:
            head_dim = len(layer.attention.q_norm_weight)
        num_heads = stack.config.num_attention_heads
        num_kv = stack.config.num_key_value_heads
        q_size = num_heads * head_dim
        kv_len = num_kv * head_dim
        layer_rope = stack.config.layer_rope_theta(layer_idx)
        layer_window = stack.config.layer_sliding_window(layer_idx)

        normed = [0.0] * (b * h)
        for t in range(b):
            scratch = [0.0] * h
            rms_norm_f32(
                hidden[t * h : (t + 1) * h],
                layer.input_layernorm,
                scratch,
                stack.config.rms_norm_eps,
            )
            normed[t * h : (t + 1) * h] = scratch

        q_all = [0.0] * (b * q_size)
        k_all = [0.0] * (b * kv_len)
        v_all = [0.0] * (b * kv_len)
        if layer.attention.q_proj.is_loaded():
            layer.attention.q_proj.gemm(normed, q_all, b)
        if layer.attention.k_proj.is_loaded():
            layer.attention.k_proj.gemm(normed, k_all, b)
        if layer.attention.v_proj.is_loaded():
            layer.attention.v_proj.gemm(normed, v_all, b)

        head_scratch = [0.0] * head_dim
        for t in range(b):
            pos = stack.position_offset + t
            q = q_all[t * q_size : (t + 1) * q_size]
            k = k_all[t * kv_len : (t + 1) * kv_len]
            if len(layer.attention.q_norm_weight) == head_dim:
                for hi in range(num_heads):
                    start = hi * head_dim
                    rms_norm_f32(
                        q[start : start + head_dim],
                        layer.attention.q_norm_weight,
                        head_scratch,
                        stack.config.rms_norm_eps,
                    )
                    q[start : start + head_dim] = head_scratch[:]
            if len(layer.attention.k_norm_weight) == head_dim:
                for hi in range(num_kv):
                    start = hi * head_dim
                    rms_norm_f32(
                        k[start : start + head_dim],
                        layer.attention.k_norm_weight,
                        head_scratch,
                        stack.config.rms_norm_eps,
                    )
                    k[start : start + head_dim] = head_scratch[:]
            for hi in range(num_heads):
                start = hi * head_dim
                apply_rope_head_f32(
                    q[start : start + head_dim],
                    head_scratch,
                    pos,
                    head_dim,
                    layer_rope,
                )
                q[start : start + head_dim] = head_scratch[:]
            for hi in range(num_kv):
                start = hi * head_dim
                apply_rope_head_f32(
                    k[start : start + head_dim],
                    head_scratch,
                    pos,
                    head_dim,
                    layer_rope,
                )
                k[start : start + head_dim] = head_scratch[:]
            # Write the normalized/roped vectors back into the batch buffers
            # (list slices above are copies, not views).
            q_all[t * q_size : (t + 1) * q_size] = q
            k_all[t * kv_len : (t + 1) * kv_len] = k

        attn_pre_o = [0.0] * (b * q_size)
        cache = stack.kv_cache[layer_idx]
        for t in range(b):
            k = k_all[t * kv_len : (t + 1) * kv_len]
            v = v_all[t * kv_len : (t + 1) * kv_len]
            cache.keys.extend(k)
            cache.values.extend(v)
            cache.seq_len += 1
            q = q_all[t * q_size : (t + 1) * q_size]
            out = [0.0] * q_size
            win_keys, win_values, win_seq = _windowed_cache(
                cache.keys, cache.values, cache.seq_len, kv_len, layer_window
            )
            flash_attention_decode_heads_gqa(
                q,
                win_keys,
                win_values,
                out,
                win_seq,
                head_dim,
                kv_len,
                num_heads,
                num_kv,
            )
            attn_pre_o[t * q_size : (t + 1) * q_size] = out

        attn_out_all = [0.0] * (b * h)
        if layer.attention.o_proj.is_loaded():
            layer.attention.o_proj.gemm(attn_pre_o, attn_out_all, b)
        elif q_size == h:
            attn_out_all[:] = attn_pre_o
        # Gemma sandwich norm: normalize attention output before residual.
        if stack.config.sandwich_norm and len(layer.post_attention_layernorm) == h:
            for t in range(b):
                scratch = [0.0] * h
                rms_norm_f32(
                    attn_out_all[t * h : (t + 1) * h],
                    layer.post_attention_layernorm,
                    scratch,
                    stack.config.rms_norm_eps,
                )
                attn_out_all[t * h : (t + 1) * h] = scratch
        for i in range(len(hidden)):
            hidden[i] += attn_out_all[i]

        mlp_norm = (
            layer.pre_ffn_layernorm
            if stack.config.sandwich_norm and layer.pre_ffn_layernorm
            else layer.post_attention_layernorm
        )
        normed_mlp = [0.0] * (b * h)
        for t in range(b):
            scratch = [0.0] * h
            rms_norm_f32(
                hidden[t * h : (t + 1) * h],
                mlp_norm,
                scratch,
                stack.config.rms_norm_eps,
            )
            normed_mlp[t * h : (t + 1) * h] = scratch
        inter = stack.config.intermediate_size
        gate = [0.0] * (b * inter)
        up = [0.0] * (b * inter)
        if layer.mlp_gate.is_loaded():
            layer.mlp_gate.gemm(normed_mlp, gate, b)
        if layer.mlp_up.is_loaded():
            layer.mlp_up.gemm(normed_mlp, up, b)
        _apply_glu(gate, up, stack.config.gelu_ffn)
        mlp_out_all = [0.0] * (b * h)
        if layer.mlp_down.is_loaded():
            layer.mlp_down.gemm(gate, mlp_out_all, b)
        # Gemma sandwich norm: normalize FFN output before residual.
        if stack.config.sandwich_norm and len(layer.post_ffn_layernorm) == h:
            for t in range(b):
                scratch = [0.0] * h
                rms_norm_f32(
                    mlp_out_all[t * h : (t + 1) * h],
                    layer.post_ffn_layernorm,
                    scratch,
                    stack.config.rms_norm_eps,
                )
                mlp_out_all[t * h : (t + 1) * h] = scratch
        for i in range(len(hidden)):
            hidden[i] += mlp_out_all[i]

    out_hidden = [0.0] * (b * h)
    for t in range(b):
        scratch = [0.0] * h
        rms_norm_f32(
            hidden[t * h : (t + 1) * h],
            stack.norm,
            scratch,
            stack.config.rms_norm_eps,
        )
        out_hidden[t * h : (t + 1) * h] = scratch
    stack.position_offset += b
    return list(out_hidden[(b - 1) * h : b * h])


def logits(stack: LlamaDecoderStack, hidden: list[float]) -> Logits:
    if not stack.output.is_loaded():
        raise ValueError("decoder stack is missing output projection")
    vocab = stack.output._output_dim()
    in_dim = stack.output._input_dim()
    if len(hidden) < in_dim:
        raise ValueError(f"hidden width {len(hidden)} smaller than output input width {in_dim}")
    out: Logits = [0.0] * vocab
    stack.output.gemv(hidden[:in_dim], out)
    return out

File: oxidize-golang/core/model/llama_decoder_forward.go
package model

import (
	"fmt"

	"github.com/Zapdev-labs/oxidize/golang/core/flash_attention"
	"github.com/Zapdev-labs/oxidize/golang/core/tensor"
)

// ForwardToken runs one decode step and returns the final hidden state (pre-lm_head).
func (s *LlamaDecoderStack) ForwardToken(token uint32) ([]float32, error) {
	return s.forwardTokenWithContext(token, nil, nil)
}

// forwardTokenWithContext runs decode with optional precomputed K/V context vectors per layer
// (used by DFlash target-hidden fusion).
func (s *LlamaDecoderStack) forwardTokenWithContext(
	token uint32,
	targetContext []float32,
	kvContext func(layerIdx int, targetCtx []float32) (kCtx, vCtx []float32, err error),
) ([]float32, error) {
	h := s.Config.HiddenSize
	hidden := make([]float32, h)

	if err := s.fillTokenEmbedding(token, hidden); err != nil {
		return nil, err
	}
	s.scaleEmbedding(hidden)

	for layerIdx := range s.Layers {
		layer := &s.Layers[layerIdx]
		headDim := s.Config.KVHeadDim()
		if len(layer.Attention.QNormWeight) > 0 {
			headDim = len(layer.Attention.QNormWeight)
		}
		numHeads := s.Config.NumAttentionHeads
		numKV := s.Config.NumKeyValueHeads
		qSize := numHeads * headDim
		kvLen := numKV * headDim

		attnOut := make([]float32, qSize)
		mlpOut := make([]float32, h)

		{
			normed := make([]float32, h)
			if err := tensor.RMSNormF32(hidden, layer.InputLayernorm, normed, s.Config.RMSNormEps); err != nil {
				return nil, err
			}

			q := make([]float32, qSize)
			k := make([]float32, kvLen)
			v := make([]float32, kvLen)
			if layer.Attention.QProj.IsLoaded() {
				if err := layer.Attention.QProj.Gemv(normed, q); err != nil {
					return nil, err
				}
			}
			if layer.Attention.KProj.IsLoaded() {
				if err := layer.Attention.KProj.Gemv(normed, k); err != nil {
					return nil, err
				}
			}
			if layer.Attention.VProj.IsLoaded() {
				if err := layer.Attention.VProj.Gemv(normed, v); err != nil {
					return nil, err
				}
			}

			var kCtx, vCtx []float32
			if kvContext != nil && targetContext != nil {
				var err error
				kCtx, vCtx, err = kvContext(layerIdx, targetContext)
				if err != nil {
					return nil, err
				}
			}

			headScratch := make([]float32, headDim)
			pos := s.PositionOffset

			applyQKNorm := func(vec []float32, weight []float32, heads int) error {
				if len(weight) != headDim {
					return nil
				}
				for hi := 0; hi < heads; hi++ {
					start := hi * headDim
					if err := tensor.RMSNormF32(vec[start:start+headDim], weight, headScratch, s.Config.RMSNormEps); err != nil {
						return err
					}
					copy(vec[start:start+headDim], headScratch)
				}
				return nil
			}
			if err := applyQKNorm(q, layer.Attention.QNormWeight, numHeads); err != nil {
				return nil, err
			}
			if err := applyQKNorm(k, layer.Attention.KNormWeight, numKV); err != nil {
				return nil, err
			}
			if kCtx != nil {
				if err := applyQKNorm(kCtx, layer.Attention.KNormWeight, numKV); err != nil {
					return nil, err
				}
			}

			layerRope := s.Config.LayerRopeTheta(layerIdx)
			if !s.Config.UseAlibi {
				applyRope := func(vec []float32, heads int) error {
					for hi := 0; hi < heads; hi++ {
						start := hi * headDim
						if err := tensor.ApplyRopeHeadF32(vec[start:start+headDim], headScratch, pos, headDim, layerRope); err != nil {
							return err
						}
						copy(vec[start:start+headDim], headScratch)
					}
					return nil
				}
				if err := applyRope(q, numHeads); err != nil {
					return nil, err
				}
				if err := applyRope(k, numKV); err != nil {
					return nil, err
				}
				if kCtx != nil {
					if err := applyRope(kCtx, numKV); err != nil {
						return nil, err
					}
				}
			}

			cache := &s.KVCache[layerIdx]
			if kCtx != nil {
				cache.Keys = append(cache.Keys, kCtx...)
				cache.Values = append(cache.Values, vCtx...)
				cache.SeqLen++
			}
			cache.Keys = append(cache.Keys, k...)
			cache.Values = append(cache.Values, v...)
			cache.SeqLen++

			window := s.Config.LayerSlidingWindow(layerIdx)
			var attnErr error
			switch {
			case s.Config.UseAlibi && len(s.AlibiSlopes) > 0:
				attnErr = flash_attention.FlashAttentionDecodeHeadsGQAAlibi(
					q, cache.Keys, cache.Values, attnOut,
					cache.SeqLen, headDim, kvLen, numHeads, numKV,
					s.AlibiSlopes, pos, window,
				)
			case window > 0:
				attnErr = flash_attention.FlashAttentionDecodeHeadsGQAWindow(
					q, cache.Keys, cache.Values, attnOut,
					cache.SeqLen, headDim, kvLen, numHeads, numKV, window,
				)
			default:
				attnErr = flash_attention.FlashAttentionDecodeHeadsGQA(
					q, cache.Keys, cache.Values, attnOut,
					cache.SeqLen, headDim, kvLen, numHeads, numKV,
				)
			}
			if attnErr != nil {
				return nil, attnErr
			}

			oResult := make([]float32, h)
			if layer.Attention.OProj.IsLoaded() {
				if err := layer.Attention.OProj.Gemv(attnOut, oResult); err != nil {
					return nil, err
				}
			} else {
				copy(oResult, attnOut)
				if qSize < h {
					for i := qSize; i < h; i++ {
						oResult[i] = 0
					}
				}
			}
			attnOut = oResult

			if s.Config.ParallelAttnFFN {
				if err := s.runPostAttentionFFN(layer, hidden, mlpOut); err != nil {
					return nil, err
				}
			}
		}

		if s.Config.ParallelAttnFFN {
			for i := 0; i < h; i++ {
				hidden[i] += attnOut[i] + mlpOut[i]
			}
		} else {
			// Gemma sandwich norm: normalize the attention output before residual.
			if s.Config.SandwichNorm && len(layer.PostAttentionLayernorm) == h {
				normedAttn := make([]float32, h)
				if err := tensor.RMSNormF32(attnOut, layer.PostAttentionLayernorm, normedAttn, s.Config.RMSNormEps); err != nil {
					return nil, err
				}
				attnOut = normedAttn
			}
			for i := 0; i < h; i++ {
				hidden[i] += attnOut[i]
			}
			if err := s.runPostAttentionFFN(layer, hidden, mlpOut); err != nil {
				return nil, err
			}
			for i := 0; i < h; i++ {
				hidden[i] += mlpOut[i]
			}
		}
	}

	if len(s.Norm) > 0 {
		out := make([]float32, h)
		if err := tensor.RMSNormF32(hidden, s.Norm, out, s.Config.RMSNormEps); err != nil {
			return nil, err
		}
		hidden = out
	}

	s.PositionOffset++
	return hidden, nil
}

// ForwardBatch prefills multiple tokens; returns the last token hidden state.
func (s *LlamaDecoderStack) ForwardBatch(tokens []uint32) ([]float32, error) {
	if len(tokens) == 0 {
		return nil, fmt.Errorf("empty token batch")
	}
	if len(tokens) == 1 {
		return s.ForwardToken(tokens[0])
	}

	b := len(tokens)
	h := s.Config.HiddenSize
	hidden := make([]float32, b*h)

	if s.TokEmbeddings.IsLoaded() {
		for t, tok := range tokens {
			if err := s.fillTokenEmbedding(tok, hidden[t*h:(t+1)*h]); err != nil {
				return nil, err
			}
			s.scaleEmbedding(hidden[t*h : (t+1)*h])
		}
	}

	for layerIdx := range s.Layers {
		layer := &s.Layers[layerIdx]
		headDim := s.Config.KVHeadDim()
		if len(layer.Attention.QNormWeight) > 0 {
			headDim = len(layer.Attention.QNormWeight)
		}
		numHeads := s.Config.NumAttentionHeads
		numKV := s.Config.NumKeyValueHeads
		qSize := numHeads * headDim
		kvLen := numKV * headDim
		layerRope := s.Config.LayerRopeTheta(layerIdx)
		layerWindow := s.Config.LayerSlidingWindow(layerIdx)

		normed := make([]float32, b*h)
		for t := 0; t < b; t++ {
			if err := tensor.RMSNormF32(hidden[t*h:(t+1)*h], layer.InputLayernorm, normed[t*h:(t+1)*h], s.Config.RMSNormEps); err != nil {
				return nil, err
			}
		}

		qAll := make([]float32, b*qSize)
		kAll := make([]float32, b*kvLen)
		vAll := make([]float32, b*kvLen)
		if layer.Attention.QProj.IsLoaded() {
			if err := layer.Attention.QProj.Gemm(normed, qAll, b); err != nil {
				return nil, err
			}
		}
		if layer.Attention.KProj.IsLoaded() {
			if err := layer.Attention.KProj.Gemm(normed, kAll, b); err != nil {
				return nil, err
			}
		}
		if layer.Attention.VProj.IsLoaded() {
			if err := layer.Attention.VProj.Gemm(normed, vAll, b); err != nil {
				return nil, err
			}
		}

		headScratch := make([]float32, headDim)
		for t := 0; t < b; t++ {
			pos := s.PositionOffset + t
			q := qAll[t*qSize : (t+1)*qSize]
			k := kAll[t*kvLen : (t+1)*kvLen]
			if len(layer.Attention.QNormWeight) == headDim {
				for hi := 0; hi < numHeads; hi++ {
					start := hi * headDim
					if err := tensor.RMSNormF32(q[start:start+headDim], layer.Attention.QNormWeight, headScratch, s.Config.RMSNormEps); err != nil {
						return nil, err
					}
					copy(q[start:start+headDim], headScratch)
				}
			}
			if len(layer.Attention.KNormWeight) == headDim {
				for hi := 0; hi < numKV; hi++ {
					start := hi * headDim
					if err := tensor.RMSNormF32(k[start:start+headDim], layer.Attention.KNormWeight, headScratch, s.Config.RMSNormEps); err != nil {
						return nil, err
					}
					copy(k[start:start+headDim], headScratch)
				}
			}
			if !s.Config.UseAlibi {
				for hi := 0; hi < numHeads; hi++ {
					start := hi * headDim
					if err := tensor.ApplyRopeHeadF32(q[start:start+headDim], headScratch, pos, headDim, layerRope); err != nil {
						return nil, err
					}
					copy(q[start:start+headDim], headScratch)
				}
				for hi := 0; hi < numKV; hi++ {
					start := hi * headDim
					if err := tensor.ApplyRopeHeadF32(k[start:start+headDim], headScratch, pos, headDim, layerRope); err != nil {
						return nil, err
					}
					copy(k[start:start+headDim], headScratch)
				}
			}
		}

		attnPreO := make([]float32, b*qSize)
		cache := &s.KVCache[layerIdx]
		for t := 0; t < b; t++ {
			k := kAll[t*kvLen : (t+1)*kvLen]
			v := vAll[t*kvLen : (t+1)*kvLen]
			cache.Keys = append(cache.Keys, k...)
			cache.Values = append(cache.Values, v...)
			cache.SeqLen++
			q := qAll[t*qSize : (t+1)*qSize]
			out := attnPreO[t*qSize : (t+1)*qSize]
			pos := s.PositionOffset + t
			switch {
			case s.Config.UseAlibi && len(s.AlibiSlopes) > 0:
				if err := flash_attention.FlashAttentionDecodeHeadsGQAAlibi(
					q, cache.Keys, cache.Values, out,
					cache.SeqLen, headDim, kvLen, numHeads, numKV,
					s.AlibiSlopes, pos, layerWindow,
				); err != nil {
					return nil, err
				}
			case layerWindow > 0:
				if err := flash_attention.FlashAttentionDecodeHeadsGQAWindow(
					q, cache.Keys, cache.Values, out,
					cache.SeqLen, headDim, kvLen, numHeads, numKV, layerWindow,
				); err != nil {
					return nil, err
				}
			default:
				if err := flash_attention.FlashAttentionDecodeHeadsGQA(
					q, cache.Keys, cache.Values, out,
					cache.SeqLen, headDim, kvLen, numHeads, numKV,
				); err != nil {
					return nil, err
				}
			}
		}

		attnOutAll := make([]float32, b*h)
		if layer.Attention.OProj.IsLoaded() {
			if err := layer.Attention.OProj.Gemm(attnPreO, attnOutAll, b); err != nil {
				return nil, err
			}
		} else if qSize == h {
			copy(attnOutAll, attnPreO)
		}
		// Gemma sandwich norm: normalize attention output before residual.
		if s.Config.SandwichNorm && len(layer.PostAttentionLayernorm) == h {
			scratch := make([]float32, h)
			for t := 0; t < b; t++ {
				slice := attnOutAll[t*h : (t+1)*h]
				if err := tensor.RMSNormF32(slice, layer.PostAttentionLayernorm, scratch, s.Config.RMSNormEps); err != nil {
					return nil, err
				}
				copy(slice, scratch)
			}
		}
		for i := range hidden {
			hidden[i] += attnOutAll[i]
		}

		mlpNorm := s.mlpNorm(layer)
		normedMLP := make([]float32, b*h)
		for t := 0; t < b; t++ {
			if err := tensor.RMSNormF32(hidden[t*h:(t+1)*h], mlpNorm, normedMLP[t*h:(t+1)*h], s.Config.RMSNormEps); err != nil {
				return nil, err
			}
		}
		inter := s.Config.IntermediateSize
		gate := make([]float32, b*inter)
		up := make([]float32, b*inter)
		if layer.MLPGate.IsLoaded() {
			if err := layer.MLPGate.Gemm(normedMLP, gate, b); err != nil {
				return nil, err
			}
		}
		if layer.MLPUp.IsLoaded() {
			if err := layer.MLPUp.Gemm(normedMLP, up, b); err != nil {
				return nil, err
			}
		}
		applyGLU(gate, up, s.Config.GeluFFN)
		mlpOutAll := make([]float32, b*h)
		if layer.MLPDown.IsLoaded() {
			if err := layer.MLPDown.Gemm(gate, mlpOutAll, b); err != nil {
				return nil, err
			}
		}
		// Gemma sandwich norm: normalize FFN output before residual.
		if s.Config.SandwichNorm && len(layer.PostFFNLayernorm) == h {
			scratch := make([]float32, h)
			for t := 0; t < b; t++ {
				slice := mlpOutAll[t*h : (t+1)*h]
				if err := tensor.RMSNormF32(slice, layer.PostFFNLayernorm, scratch, s.Config.RMSNormEps); err != nil {
					return nil, err
				}
				copy(slice, scratch)
			}
		}
		for i := range hidden {
			hidden[i] += mlpOutAll[i]
		}
	}

	outHidden := make([]float32, b*h)
	for t := 0; t < b; t++ {
		if err := tensor.RMSNormF32(hidden[t*h:(t+1)*h], s.Norm, outHidden[t*h:(t+1)*h], s.Config.RMSNormEps); err != nil {
			return nil, err
		}
	}
	s.PositionOffset += b
	last := outHidden[(b-1)*h : b*h]
	return append([]float32(nil), last...), nil
}

// Logits projects hidden state to vocabulary logits.
func (s *LlamaDecoderStack) Logits(hidden []float32) (Logits, error) {
	if !s.Output.IsLoaded() {
		return nil, fmt.Errorf("decoder stack is missing output projection")
	}
	vocab := s.Output.outputDim()
	inDim := s.Output.inputDim()
	if len(hidden) < inDim {
		return nil, fmt.Errorf("hidden width %d smaller than output input width %d", len(hidden), inDim)
	}
	logits := make(Logits, vocab)
	if err := s.Output.Gemv(hidden[:inDim], logits); err != nil {
		return nil, err
	}
	return logits, nil
}
